sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala

```diff
@@ -76,41 +76,43 @@ object DataSourceV2Strategy extends Strategy {
   /**
    * Applies column pruning to the data source, w.r.t. the references of the given expressions.
    *
-   * @return new output attributes after column pruning.
+   * @return the created `ScanConfig`(since column pruning is the last step of operator pushdown),
+   *         and new output attributes after column pruning.
```

Comment:
Technically column pruning is not operator pushdown. It's kind of a property of the SQL operator, indicating which columns/nested fields the parent operator needs.

Column pruning is orthogonal to any operator pushdown: when operator pushdown is finished, Spark will check the remaining query plans at Spark side(assuming some predicates are not pushed), calculate the required columns for the scan node, and pass it as a hint to the data source.

I'm fine with the current API, but we should have a better API for column pruning in the future.

BTW for now I think adding `prunedSchema` to `SupportsPushdownRequiredColumns` is better. Data sources that don't support operator pushdown don't need to implement `ScanConfig.readSchema`. But I don't have a strong opinion.

Reply:
Why is column pruning not "technically" operator pushdown? This is done by pushing a Project operator to the source.

Reply:
no it's not a project operator. If we want to support project pushdown (like `a + b`) in the future, it must be a different API. For project pushdown, operators order does matter.

Reply:
Okay, I see what you mean that we can't currently push down everything that is done by a projection. I'm not sure how relevant that point is, though. The implementation is still allowed to project more columns than Spark requests based on the other operations that are pushed.